src/codegen/encoding/mod.rs

```diff
@@ -0,0 +1,580 @@
+mod borsh_encoding;
+
+use crate::ast::{Namespace, RetrieveType, Type};
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
+use crate::codegen::encoding::borsh_encoding::BorshEncoding;
+use crate::codegen::expression::load_storage;
+use crate::codegen::vartable::Vartable;
+use crate::codegen::{Builtin, Expression};
+use crate::Target;
+use num_bigint::BigInt;
+use solang_parser::pt::Loc;
+
+/// This trait should be implemented by all encoding methods (ethabi, Scale and Borsh), so that
+/// we have the same interface for creating encode and decode functions.
+pub(super) trait Encoding {
+    /// Receive the arguments and returns the variable containing a byte array
+    fn abi_encode(
+        &mut self,
+        loc: &Loc,
+        args: &[Expression],
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> Expression;
+
+    /// Cache items loaded from storage to reuse them later
+    fn cache_storage_loaded(&mut self, arg_no: usize, expr: Expression);
+}
+
+/// This function should return the correct encoder, given the target
+pub(super) fn create_encoder(ns: &Namespace) -> impl Encoding {
+    match &ns.target {
+        Target::Solana => BorshEncoding::new(),
+        _ => unreachable!("Other types of encoding have not been implemented yet"),
+    }
+}
+
+/// Calculate the size of a set of arguments to encoding functions
+fn calculate_size_args<T: Encoding>(
```

Comment:
With scale encoding, we will only estimate the size (or minimum size). I guess this is a future problem.